internal/server/data/migrations.go

```diff
@@ -594,7 +596,124 @@ func setDefaultOrgID() *migrator.Migration {
 				}
 			}
 			return nil
+		},
+	}
+}
+
+func addProviderGroups() *migrator.Migration {
+	return &migrator.Migration{
+		ID: "2022-09-07T17:05",
+		Migrate: func(tx migrator.DB) error {
+			// setup the new tables
+			_, err := tx.Exec(
+				`
+					CREATE TABLE IF NOT EXISTS provider_groups (
+						id bigint NOT NULL,
+						created_at timestamp with time zone,
+						updated_at timestamp with time zone,
+						deleted_at timestamp with time zone,
+						provider_id bigint NOT NULL,
+						name text NOT NULL,
+						organization_id bigint
+					);
+					CREATE TABLE IF NOT EXISTS provider_groups_identities (
+						provider_group_id bigint NOT NULL,
+						identity_id bigint NOT NULL
+					);
+					CREATE TABLE IF NOT EXISTS provider_groups_groups (
+						provider_group_id bigint NOT NULL,
+						group_id bigint NOT NULL
+					);
+					CREATE UNIQUE INDEX idx_org_provider_group_names ON provider_groups(organization_id, provider_id, name);
+			`)
+			if err != nil {
+				return fmt.Errorf("create provider groups table: %w", err)
+			}
 
+			if migrator.HasColumn(tx, "provider_users", "groups") {
+				// move the old data into the new tables
+				stmt := `
+					SELECT identity_id, provider_id, groups, identities.organization_id 
+					FROM provider_users
+					JOIN identities ON identities.id = provider_users.identity_id
+					WHERE groups != ''
+				`
+				rows, err := tx.Query(stmt)
+				if err != nil {
+					return fmt.Errorf("migrate existing provider groups: %w", err)
+				}
+
+				type groupProviderUser struct {
+					IdentityID uid.ID
+					ProviderID uid.ID
+					Groups     models.CommaSeparatedStrings
+					OrgID      uid.ID
+				}
+				for rows.Next() {
+					var groupProviderUser groupProviderUser
+					if err := rows.Scan(&groupProviderUser.IdentityID, &groupProviderUser.ProviderID, &groupProviderUser.Groups, &groupProviderUser.OrgID); err != nil {
+						return err
+					}
+
+					for _, name := range groupProviderUser.Groups {
+						// check if this provider user group exists yet
+						var providerGroupID uid.ID
+						err := tx.QueryRow(`SELECT id FROM provider_groups WHERE name = ? AND provider_id = ?`, name, groupProviderUser.ProviderID).Scan(&providerGroupID)
+						if err != nil && !strings.Contains(err.Error(), "no rows in result set") {
+							return err
+						}
+
+						if providerGroupID == 0 {
+							// get the corresponding infra group which already exists
+							var infraGroupID uid.ID
+							err := tx.QueryRow(`SELECT id FROM groups WHERE name = ? AND organization_id = ?`, name, groupProviderUser.OrgID).Scan(&infraGroupID)
+							if err != nil {
+								return err
+							}
+
+							// create the provider group, the infra group already exists
```

Comment:
shouldn't the provider group be created regardless? it's a copy of what the provider says the groups are.

Reply:
I think this is a confusing comment, that is what is happening. I meant that the "parent" `Group` will already exist for each comma separated value on a `ProviderUser`. Compared to in the identity sync code where the `Group` might not exist.

Reply:
ok I'm not sure that clarified it for me. :D